Parallelize createOwnerInformation

atotoun opened the pull request:
`ReceivedPartition::createOwnerInformation`: the function is parallelized for bounding-box initialization according to [the issue 633](https://github.com/precice/precice/issues/633)

uekerman commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -124,6 +126,14 @@ void ReceivedPartition::compute()
   if (_toMapping)
     _toMapping->tagMeshFirstRound();
 
+  // if(utils::MasterSlave::getRank()==0)
+  // {
+  //   for (auto & vtx : _mesh->vertices())
+  //   {
+  //     std::cout<< vtx.getGlobalIndex() << std::endl;
+  //   }
+  // }
```
Let's not push this to develop. If you think this is helpful in the long run you could make it a logging output and potentially guard with `#ifndef NDEBUG`.

uekerman commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -161,6 +171,14 @@ void ReceivedPartition::compute()
     // A vertex belongs to a specific connected rank if its global vertex ID lies within the ranks min and max.
     std::map<int, std::vector<int>> remoteCommunicationMap;
 
+    // if(utils::MasterSlave::getRank()==0)
+    // {
+    //   for (auto & vtx : _mesh->vertices())
+    //   {
+    //     std::cout<< vtx.getGlobalIndex() << std::endl;
+    //   }
+    // }
```
Let's not push this to develop. If you think this is helpful in the long run you could make it a logging output and potentially guard with `#ifndef NDEBUG`.

uekerman commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -470,143 +488,312 @@ void ReceivedPartition::createOwnerInformation()
+
+    // #1: receive local bb map from master
+    // Define and initialize localBBMap to save local bbs
+    prepareBoundingBox();
```
Shouldn't that already happen in `compareBoundingBoxes` ?

atotoun replied:
That's right.

uekerman commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -470,143 +488,312 @@ void ReceivedPartition::createOwnerInformation()
+
+    if (utils::MasterSlave::isMaster()) {
+
+      // master receives local bb from each slave ranke
```
```suggestion
      // master receives local bb from each slave rank
```

uekerman commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -470,143 +488,312 @@ void ReceivedPartition::createOwnerInformation()
+        com::CommunicateBoundingBox(utils::MasterSlave::_communication).receiveBoundingBox(localBBMap.at(rankSlave), rankSlave);
+      }
+
+      // master broadcast localBBMap to all slaves 
+      for (int rankSlave = 1; rankSlave < utils::MasterSlave::getSize(); rankSlave++) {
+        com::CommunicateBoundingBox(utils::MasterSlave::_communication).sendBoundingBoxMap(localBBMap, rankSlave);
+      }
```
It would performance-wise be better to use a proper broadcast here.

atotoun replied:
True, I replaced it.

uekerman commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -470,143 +488,312 @@ void ReceivedPartition::createOwnerInformation()
 
-  if (utils::MasterSlave::isSlave()) {
+  if (m2n().usesTwoLevelInitialization()) {
+    /*
+    1- receive local bb map from master
+    2- filter bb map to keep the connected ranks
+    3- own the vertices that only fits into the ranks bb
+    4- send number of owned vertices and shared vertices list to neighbors
+    5- for the remaining vertices: check if we have less vertexes -> own it!
+    */
```
Could you please additionally add a few lines of big-picture explanation here. What is done and why.
The steps you have are already very helpful, but still require the reader to understand how you got there.

atotoun replied:
Explanation is added. Please check for correctness and completeness.

uekerman commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -470,143 +488,312 @@ void ReceivedPartition::createOwnerInformation()
+    }    
+
+    // #2: filter bb map to keep the connected ranks
+    // remove the current bb from the map
```
```suggestion
    // remove the own bb from the map
```
please add an explanation why we do this

uekerman commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -470,143 +488,312 @@ void ReceivedPartition::createOwnerInformation()
+    int receivedCounter = 0;
+    for (auto & neighborRank: localConnectedBBMap)
+    {
+      utils::MasterSlave::_communication->aSend(counter, neighborRank.first);
```
I don't like the `aSend` here a lot. I guess it could still lead to a deadlock for very large cases depending on the MPI implementation.
Better could be:
* first `aReceive`
* then `send`
* last, check that all receives did happen

atotoun replied:
I don't understand how this could result in a deadlock. Maybe I can check after aSend? 

I think what you suggest won't work, since here we first need to exchange size and if not zero communicate the vertex list. 
aReceive for size will always result in 0 size and no aReceive for vertex list will be called.

atotoun replied:
I replaced aSend/receive with a aReceive/send. 

For the rest, let's continue the discussion below.

uekerman commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -470,143 +488,312 @@ void ReceivedPartition::createOwnerInformation()
+      }
+    }
+
+    // #3: filter vertices and keep those only fit into the current rank's bb
```
Let's not use the word filter here. In our context, filter means that we delete vertices, which we do not do here.

uekerman commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -470,143 +488,312 @@ void ReceivedPartition::createOwnerInformation()
+      for (auto & sharingRank : sharedVerticesReceiveMap)
+      {
+        std::vector<double> vec = sharingRank.second;
+        if( std::find(vec.begin(), vec.end(), sharedVerticesList[i]) != vec.end() )
+        {
+          if((counter > neighborRanksVertexCount[sharingRank.first]) ||
+             (counter == neighborRanksVertexCount[sharingRank.first] && utils::MasterSlave::getRank() > sharingRank.first) )
+          {
+            owned = false;
+          }
         }
```
Could you please add some documentation on how you do the load balancing?
Why the `utils::MasterSlave::getRank() > sharingRank.first`?

atotoun replied:
Documentation is added. 
This is for the case where a vertex can be shared between two ranks and they have same vertex count: the lower rank will own the vertex.

uekerman commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -470,143 +488,312 @@ void ReceivedPartition::createOwnerInformation()
+    // end of bb section 
+  } else
+  {
+    std::cout<<"we use normal filtering ..." << std::endl;
```
Please use Debug log

uekerman commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -470,143 +488,312 @@ void ReceivedPartition::createOwnerInformation()
-                                    << " vertices of mesh " << _mesh->getName() << " have been filtered out "
-                                    << "since they have no influence on the mapping.");
+      auto filteredVertices = std::count(globalOwnerVec.begin(), globalOwnerVec.end(), 0);
+      if (filteredVertices)
+        PRECICE_WARN(filteredVertices << " of " << _mesh->getGlobalNumberOfVertices()
+                     << " vertices of mesh " << _mesh->getName() << " have been filtered out "
+                     << "since they have no influence on the mapping.");
+    }
```
I guess those could be helpful for the two-level variant as well.

atotoun replied:
Agreed and added.

uekerman reviewed (commented):
I think this goes in the right direction :+1: Maybe some things could still be simplified.
I have not yet checked all details and have not yet looked at the tests.

atotoun commented:
@uekerman  I applied all the comments except one which I do not completely understand. Could you please have a look and see if you agree with my reply?

uekerman commented on src/partition/tests/ReceivedPartitionTest.cpp:
```diff
@@ -1087,6 +1087,147 @@ BOOST_AUTO_TEST_CASE(TestCompareBoundingBoxes3D)
+      BOOST_TEST(mesh->getCommunicationMap()[0][1] == 1);
+      BOOST_TEST(mesh->getCommunicationMap()[0][2] == 2);
+
+      // std::cout<<"master sends vertices: " << mesh->getCommunicationMap()[0] << " to rank 0 and verticees " << mesh->getCommunicationMap()[1] << " to rank 1 " << std::endl; 
```
Let's not merge these.

uekerman commented on src/partition/tests/ReceivedPartitionTest.cpp:
```diff
@@ -1087,6 +1087,147 @@ BOOST_AUTO_TEST_CASE(TestCompareBoundingBoxes3D)
+    m2n->requestSlavesPreConnection("FluidSlaves", "SolidSlaves");
+
+    part.communicate();
+    part.compute();
```
Shouldn't we test whether the actual owner information is as expected?

atotoun replied:
Right, test is modified.

uekerman reviewed (changes requested):
I had a more thorough look. In principle the algorithmic idea is nice, but I am not yet sure if it does always work. We rely on the fact that all ranks always make the same decision on who is the owner, right? At the same time the decision depends on "counter", which changes dynamically. A particularly complicated case is if three ranks could own a vertex.
We definitely need to test more edge cases.

atotoun commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -464,145 +466,330 @@ void ReceivedPartition::createOwnerInformation()
+    */
+
+    // #1: receive local bb map from master
+    // Define and initialize localBBMap to save local bbs
```
Since we also communicate the remote BBs in other sections, keeping local would help to avoid confusion. I would say let's keep it.

atotoun commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -464,145 +466,330 @@ void ReceivedPartition::createOwnerInformation()
+             (counter == neighborRanksVertexCount[sharingRank.first] && utils::MasterSlave::getRank() > sharingRank.first) )
+          {
+            owned = false;
+          }
```
Yes, that's right. 
If a vertex is owned by n ranks, this means the vertex is located in the BB of all of them -> all n ranks are neighbors and have intersecting BBs -> each rank compares the counter against all other n-1 -> this results in a unique decision. 

I agree with the edge testing, I all come up with some thing.

uekerman replied:
Yes, I think this can work. Before I thought that `counter` gets increased during this "2nd round", but it is actually frozen, which I guess is OK. Should not result in too much load imbalance.

atotoun commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -464,145 +466,330 @@ void ReceivedPartition::createOwnerInformation()
+      {
+        std::vector<double> receivedSharedVertices;
+        utils::MasterSlave::_communication->receive(receivedSharedVertices, neighborRank.first);
+        sharedVerticesReceiveMap.insert(std::make_pair(neighborRank.first, receivedSharedVertices));
```
I still do not completely understand why aSend/receive can result in a deadlock but send/aReceive won't. 

The problem with aReceive/send is, that we have to separate the communication of number of vertices and vertices list. This is because, we first need to check if there is any vertex to communicate. If we do them together, the list size would always be 0, since it takes time to aReceive gets it,  and therefore, it will not enter to the if where it receives the actual list. 

If we separate the size and list communication, we would need to define another map to save the message size for each rank separately. This would be ugly in my opinion, and I would avoid it unless we make sure that the deadlock threat is real ;)  

Regarding the check after aSend, I think it is not necessary, since the aSend is followed by a receive. This is an implicit barrier, right?

uekerman replied:
aSend uses `MPI_Isend` under the hood. There is no guarantee that this function is asynchronous. For large messages (such as the list of shared vertices at some point?) it might block. More information [here](https://www.mcs.anl.gov/research/projects/mpi/sendmode.html) for example.
Whenever we use an `aSend` or an `aReceive`, we should always check the request. Otherwise this is prone to fail if somebody else further develops this code part. You cannot be sure that the implicit barrier will always be there. Also, MPI does not give any guarantee that this implicit barrier is enough.
If we don't need the request, we could also just use a normal `send`.

atotoun commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -464,145 +466,330 @@ void ReceivedPartition::createOwnerInformation()
+    mesh::Mesh::BoundingBoxMap localBBMap;
+    for (int rank = 0; rank < utils::MasterSlave::getSize(); rank++) {
+      localBBMap.emplace(rank, _bb);
+    }
```
changed.

atotoun commented:
All comments are considered, either applied or replied. @uekerman ready for check.

uekerman commented:
Looks good in principle. 
* I still think it would be good to wait till the asynchronous communication succeeds.
* Moreover, I have a bit a bad feeling about the test coverage. We could actually do much simpler tests here. No need to have a "other participant". You could just create a `ReceivedPartition` on one participant and really unit test `createOwnerInformation()`. By creating a general test hull (similar to [PetRBF](https://github.com/precice/precice/blob/f1cb45f92cd5148174461d6123c69b2f2c92d08b/src/mapping/tests/PetRadialBasisFctMappingTest.cpp#L126) for example) and then probing it with different data sets. Understandable what I mean?

uekerman commented on src/partition/tests/ReceivedPartitionTest.cpp:
```diff
@@ -1071,414 +1074,330 @@ BOOST_AUTO_TEST_CASE(TestParallelSetOwnerInformation2D)
   double        safetyFactor = 0;
   mesh::PtrMesh mesh(new mesh::Mesh("mesh", dimensions, testing::nextMeshID()));
-  mesh::PtrMesh receivedMesh(new mesh::Mesh("mesh", dimensions, testing::nextMeshID()));
+  // mesh::PtrMesh receivedMesh(new mesh::Mesh("mesh", dimensions, testing::nextMeshID()));
```
```suggestion
```

uekerman reviewed (commented):
The tests still only cover very simple cases: 

* If shared vertices, all ranks hold them
* Then always rank 0 is owner
* No cases where the number of already owned vertices is used as criterion
* No cases with empty partitions

 With a general test hull, we could increase the coverage significantly without duplicating much code.

atotoun commented:
In fact, the first two cases (lower rank and rank with already lower vertices) had already been covered. I refactored the tests with the general hull and added a test for the case with empty partition.

uekerman commented on src/partition/tests/ReceivedPartitionTest.cpp:
```diff
@@ -1066,6 +1069,415 @@ BOOST_AUTO_TEST_CASE(TestCompareBoundingBoxes3D)
+    position << -2.0, 1.0, 1.0;
+    mesh->createVertex(position);
+  } else if (context.isRank(2)) {
+    Eigen::VectorXd position(dimensions);
```
```suggestion
```

uekerman commented on src/partition/tests/ReceivedPartitionTest.cpp:
```diff
@@ -1066,6 +1069,415 @@ BOOST_AUTO_TEST_CASE(TestCompareBoundingBoxes3D)
+{
+  /*
+    This test examines an edge case for parallel setOwnerinformation function in receivedpartition.cpp
+    for 2LI. The provided mesh includes a vertices at point (0, 0, 0) and (0, 0, 1). The rank 2 has an 
```
```suggestion
    for 2LI. The provided mesh includes vertices at points (0, 0, 0) and (0, 0, 1). Rank 2 has an 
```

uekerman reviewed (approved):
Test coverage looks better now :+1:
I guess, we need the templating  of `mesh` for the friend decleration of `testParallelSetOwnerInformation`?

@fsimonis What's the best practice here for white box testing? Can we avoid the templating?

atotoun commented:
@uekerman Yes, the compiler does not recognize the mesh at this point. I guess one way would be forward declaration of mesh class before this, but I am not sure how this would affect the rest of the things. Using Template seemed the easiest and cleanest to me. Open to better suggestions @fsimonis.

fsimonis commented:
Hi, I think the easiest would be to use something analogue to the WhiteBoxAccessor and the `impl()` function.
* Add a new struct
* Make it a friend of whatever class X it needs to access.
* Then add a function that given an &X, returns what you want to inspect.
* Then create a Boost test using a the accessor as Fixture.

https://github.com/precice/precice/blob/f71e555ca03c6b0f11443088a3b651b6e747033d/src/testing/Testing.hpp#L47-L62

https://github.com/precice/precice/blob/f71e555ca03c6b0f11443088a3b651b6e747033d/src/precice/SolverInterface.hpp#L787-L788

https://github.com/precice/precice/blob/f71e555ca03c6b0f11443088a3b651b6e747033d/src/precice/SolverInterface.hpp#L15-L17

Docs for using [Boost.Test with fixtures](https://www.boost.org/doc/libs/1_60_0/libs/test/doc/html/boost_test/tests_organization/fixtures/case.html). You need to use the accessor here, then you can directly call its member functions in the test.

atotoun commented on src/partition/tests/ReceivedPartitionTest.cpp:
```diff
@@ -1066,6 +1069,414 @@ BOOST_AUTO_TEST_CASE(TestCompareBoundingBoxes3D)
+  bool   flipNormals  = true;
+  double safetyFactor = 0;
+
+  testing::ConnectionOptions options;
+  options.useOnlyMasterCom = false;
+  options.useTwoLevelInit  = true;
+  options.type             = testing::ConnectionType::PointToPoint;
+
+  auto                                      participantCom = com::PtrCommunication(new com::SocketCommunication());
+  m2n::DistributedComFactory::SharedPointer distrFactory;
+
+  auto m2n = m2n::PtrM2N(new m2n::M2N(participantCom, distrFactory, options.useOnlyMasterCom, options.useTwoLevelInit));
```
Since there is only one participant, this does not work. It fails when connecting masters, since there is only one master.

fsimonis replied:
Ah you are right! Sorry about that

atotoun commented on src/partition/tests/ReceivedPartitionTest.cpp:
```diff
@@ -1066,6 +1069,414 @@ BOOST_AUTO_TEST_CASE(TestCompareBoundingBoxes3D)
   tearDownParallelEnvironment();
 }
 
+template <typename T>
+void testParallelSetOwnerInformation(T &mesh, int dimensions)
```
Doing: 
```
void testParallelSetOwnerInformation(mesh::PtrMesh &mesh, int dimensions)
```

fsimonis replied:
Yes, you need to change the friend declaration too.

atotoun replied:
That way compilation fails. Compiler does not recognize mesh at this point. 
No need anymore, I implemented the fixture, no test friendship needed!

atotoun commented:
> Hi, I think the easiest would be to use something analogue to the WhiteBoxAccessor and the `impl()` function.
> 
> * Add a new struct
> * Make it a friend of whatever class X it needs to access.
> * Then add a function that given an &X, returns what you want to inspect.
> * Then create a Boost test using a the accessor as Fixture.
> 
> https://github.com/precice/precice/blob/f71e555ca03c6b0f11443088a3b651b6e747033d/src/testing/Testing.hpp#L47-L62
> 
> https://github.com/precice/precice/blob/f71e555ca03c6b0f11443088a3b651b6e747033d/src/precice/SolverInterface.hpp#L787-L788
> 
> https://github.com/precice/precice/blob/f71e555ca03c6b0f11443088a3b651b6e747033d/src/precice/SolverInterface.hpp#L15-L17
> 
> Docs for using [Boost.Test with fixtures](https://www.boost.org/doc/libs/1_60_0/libs/test/doc/html/boost_test/tests_organization/fixtures/case.html). You need to use the accessor here, then you can directly call its member functions in the test.

I agree that adding a whightboxaccessor would help for further testing, as we would have this as a friend. However, in this test we really need to make the test friend of the class, as we want to call some private functions of this class. We already have access to what we want to inspect (mesh object), as we pass mesh by reference.

fsimonis commented on src/partition/tests/ReceivedPartitionTest.cpp:
```diff
@@ -1066,6 +1069,414 @@ BOOST_AUTO_TEST_CASE(TestCompareBoundingBoxes3D)
+  part.addFromMapping(boundingFromMapping);
+  part.addToMapping(boundingToMapping);
+
+  part._mesh->computeBoundingBox();
```
Can't you call this directly on the mesh instead going through the partition?

atotoun replied:
yes, I can
done!

fsimonis commented:
> I agree that adding a whightboxaccessor would help for further testing, as we would have this as a friend. However, in this test we really need to make the test friend of the class, as we want to directly call some private functions of this class. We already have access to what we want to inspect (mesh object), as we pass mesh by reference.

If you need to call private functions from the test, then you can also use a friend-fxture. If you want to call P.functionA(a,b,c). Then you can create add a member function to the fixture which takes the partition and the arguments and calls the private function like this:
 ```cpp
class P {
private:
 void functionA(a,b,c);
 struct PFixture;
 friend PFixture;
};

struct PFixture {
  void functionA(P& p, A& a, B& b, C& c) {
    p.functionA(a, b, c)
  }
};
```

atotoun commented:
> > I agree that adding a whightboxaccessor would help for further testing, as we would have this as a friend. However, in this test we really need to make the test friend of the class, as we want to directly call some private functions of this class. We already have access to what we want to inspect (mesh object), as we pass mesh by reference.
> 
> If you need to call private functions from the test, then you can also use a friend-fxture. If you want to call P.functionA(a,b,c). Then you can create add a member function to the fixture which takes the partition and the arguments and calls the private function like this:
> 
> ```c++
> class P {
> private:
>  void functionA(a,b,c);
>  struct PFixture;
>  friend PFixture;
> };
> 
> struct PFixture {
>  void functionA(P& p, A& a, B& b, C& c) {
>    p.functionA(a, b, c)
>  }
> };
> ```

Done.

fsimonis commented on src/partition/ReceivedPartition.hpp:
```diff
@@ -104,8 +99,28 @@ class ReceivedPartition : public Partition {
+  friend struct receivedPartitionFixture;
+};
+
+/*
+ * @brief A fixture that is used to access private functions of the receivedPartition class.
+ *
+ * The fixture can be used to call private functions for individual testing. 
+ */
+struct receivedPartitionFixture {
+  void createOwnerInformation(ReceivedPartition &part)
+  {
+    part.createOwnerInformation();
+  }
+  void tagMeshFirstRound(ReceivedPartition &part)
+  {
+    part.tagMeshFirstRound();
+  }
+  void prepareBoundingBox(ReceivedPartition &part)
+  {
+    part.prepareBoundingBox();
+  }
```
Moving this into a dedicated testing header would make things cleaner. Maybe `src/partition/tests/fixtures.hpp`?

atotoun replied:
Done.

fsimonis commented on src/partition/ReceivedPartition.hpp:
```diff
@@ -104,8 +99,28 @@ class ReceivedPartition : public Partition {
+ * The fixture can be used to call private functions for individual testing. 
+ */
+struct receivedPartitionFixture {
+  void createOwnerInformation(ReceivedPartition &part)
```
Making these functions static will allow to call the functions without creating an instance of the fixture first.

```suggestion
  static void createOwnerInformation(ReceivedPartition &part)
```

atotoun replied:
when defined static, it is not able to call class members and it throws error.

atotoun replied:
done

fsimonis commented on src/partition/tests/ReceivedPartitionTest.cpp:
```diff
@@ -1096,11 +1095,12 @@ void testParallelSetOwnerInformation(T &mesh, int dimensions)
 
-  part.tagMeshFirstRound();
-  part.createOwnerInformation();
+  receivedPartitionFixture fixture;
+  fixture.prepareBoundingBox(part);
+  fixture.tagMeshFirstRound(part);
+  fixture.createOwnerInformation(part);
```
If the functions are static one can do this:

```suggestion
  using Access = receivedPartitionFixture;
  Access::prepareBoundingBox(part);
  Access::tagMeshFirstRound(part);
  Access::createOwnerInformation(part);
```

atotoun replied:
done

fsimonis commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -469,144 +470,345 @@ void ReceivedPartition::createOwnerInformation()
+
+    for (auto &receivingRank : sharedVerticesSendMap) {
+      int  sendSize = receivingRank.second.size();
+      auto request  = utils::MasterSlave::_communication->aSend(sendSize, receivingRank.first);
+      vertexListRequests.push_back(request);
+      if (sendSize != 0) {
+        auto request = utils::MasterSlave::_communication->aSend(receivingRank.second, receivingRank.first);
+        vertexListRequests.push_back(request);
       }
```
This logic can be implemented in `com::Communication` (the base class).
We could use an additional `out` iterator to save the requests:

Something along the lines of
```cpp
template<typename OutIter>
void Communication::aSend(const std::vector<int>& v , int rank, OutIter out) {
  //  ...
  *(out++) = sSend(v.size(), receivingRank.first);
  // ...
  *(out++) = aSend(v.data(), v.size(), receivingRank.first);
  // ---
}


utils::MasterSlave::_communication->aSend(receivingRank.second, receivingRank.first, std::back_inserter(vertexListRequests));
```

atotoun replied:
I agree in general, but let's do it after pull request.

fsimonis commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -469,144 +470,345 @@ void ReceivedPartition::createOwnerInformation()
-        }
+    for (auto &neighborRank : sharedVerticesSendMap) {
+      int receiveSize = 0;
+      utils::MasterSlave::_communication->receive(receiveSize, neighborRank.first);
+      if (receiveSize != 0) {
```
Isn't this already covered by the vector overload of receive?

atotoun replied:
You mean the message size? I think not, we need to check to avoid zero-sized messages.

fsimonis commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -469,144 +470,345 @@ void ReceivedPartition::createOwnerInformation()
-        if (globalOwnerVec[slaveGlobalIDs[rank][i]] == 0 && slaveTags[rank][i] == 1) {
-          slaveOwnerVecs[rank][i]                 = 1;
-          globalOwnerVec[slaveGlobalIDs[rank][i]] = rank + 1;
+    // wait until aSend is complete.
```
```suggestion
    // wait until all aSend are complete.
```

fsimonis commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -469,144 +470,345 @@ void ReceivedPartition::createOwnerInformation()
+    2- filter bb map to keep the connected ranks
+    3- own the vertices that only fit into this rank's bb
+    4- send number of owned vertices and the list of shared vertices to neighbors
+    5- for the remaining vertices: check if we have less vertices -> own it!
```
6. is missing here.

atotoun replied:
revised.

fsimonis commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -469,144 +470,345 @@ void ReceivedPartition::createOwnerInformation()
+          }
         }
       }
+      if (owned) {
+        tags[sharedVerticesLocalIDs[i]] = 1;
+      } else {
+        tags[sharedVerticesLocalIDs[i]] = 0;
+      }
```
Can be written a bit more compact.

```suggestion
      tags[sharedVerticesLocalIDs[i]] = owned ? 1 : 0;
```

fsimonis commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -469,144 +470,345 @@ void ReceivedPartition::createOwnerInformation()
+      bool owned = true;
+
+      for (auto &sharingRank : sharedVerticesReceiveMap) {
+        std::vector<double> vec = sharingRank.second;
```
Shouldn't this be `int` instead of `double`?

atotoun replied:
right, changes.

fsimonis commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -469,144 +470,345 @@ void ReceivedPartition::createOwnerInformation()
+        if (std::find(vec.begin(), vec.end(), sharedVerticesGlobalIDs[i]) != vec.end()) {
+          if ((counter > neighborRanksVertexCount[sharingRank.first]) ||
+              (counter == neighborRanksVertexCount[sharingRank.first] && utils::MasterSlave::getRank() > sharingRank.first)) {
+            owned = false;
```
At this moment `owned` won't change anymore.
So you could use this loop body as a predicate and  use [`std::any_of`](https://en.cppreference.com/w/cpp/algorithm/all_any_none_of) on  `sharedVerticesReceiveMap` or simply `break` to leave the for loop.

atotoun replied:
done.

atotoun commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -469,144 +470,345 @@ void ReceivedPartition::createOwnerInformation()
+    // Define and initialize localBBMap to save local bbs
+
+    mesh::Mesh::BoundingBoxMap localBBMap;
+    mesh::BoundingBox          dummyBB(_dimensions);
+    for (int rank = 0; rank < utils::MasterSlave::getSize(); rank++) {
+      localBBMap.emplace(rank, dummyBB);
```
done

atotoun commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -469,144 +470,345 @@ void ReceivedPartition::createOwnerInformation()
+    std::vector<int> sharedVerticesLocalIDs;
+
+    // store possible shared vertices in a map to communicate with neighbors, map: rank -> vertex_global_id
+    std::map<int, std::vector<double>> sharedVerticesSendMap;
+
+    // receive list of possible shared vertices from neighboring ranks
+    std::map<int, std::vector<double>> sharedVerticesReceiveMap;
```
That's right, changed. I also added necessary asend functions for vector<int> communication in com class.

atotoun commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -469,144 +470,345 @@ void ReceivedPartition::createOwnerInformation()
+    std::vector<int> tags(numberOfVertices, -1);
+    std::vector<int> globalIDs(numberOfVertices, -1);
+    bool             atInterface = false;
+    int              counter     = 0; // number of vertices owned by this rank
```
changes counter to ownedVerticesCount. 

I need the loop index i inside the loop, thus I avoided using const auto ...

atotoun commented on src/partition/ReceivedPartition.cpp:
```diff
@@ -469,144 +470,345 @@ void ReceivedPartition::createOwnerInformation()
+    std::map<int, std::vector<double>> sharedVerticesSendMap;
+
+    // receive list of possible shared vertices from neighboring ranks
+    std::map<int, std::vector<double>> sharedVerticesReceiveMap;
```
done

BenjaminRodenberg commented on src/partition/ReceivedPartition.hpp:
```diff
@@ -95,6 +98,9 @@ class ReceivedPartition : public Partition {
   /// Min global vertex IDs of remote connected ranks
   std::vector<int> _remoteMinGlobalVertexIDs;
+
+  /// Make the fixture friend of this class
+  friend struct receivedPartitionFixture;
```
I'm just using the same pattern in #1050. I think the `friend` declaration should go before the access modifiers. Then you also directly see the friends. This looks more reasonable to me, since a private or public friend is a bit odd (even though I believe it defaults to private?). I would suggest to follow this structure:
```
class A {
friend struct friendOfA;
public:
    ...
private:
    ...
};
```

I found a similar code snipped here: https://docs.microsoft.com/en-us/cpp/cpp/friend-cpp?view=msvc-160.

However, there is also the structure you are using shown in cppreference. See https://en.cppreference.com/w/cpp/language/friend. In the end it seems to be a matter of taste. Any opinions?

atotoun replied:
Yes, it seems more natural to me as well. I changed the declaration location as you suggest. 
One more thing: I separate the declaration and implementation of the fixture. I use a separate hpp file for that, which I think helps readability.

BenjaminRodenberg commented on src/partition/ReceivedPartition.hpp:
```diff
@@ -23,6 +23,9 @@ namespace partition {
  */
 class ReceivedPartition : public Partition {
+  /// Make the fixture friend of this class
+  friend struct receivedPartitionFixture;
```
```suggestion
  friend struct ReceivedPartitionFixture;
```

atotoun replied:
done

BenjaminRodenberg commented on src/partition/tests/ReceivedPartitionTest.cpp:
```diff
@@ -1068,6 +1072,414 @@ BOOST_AUTO_TEST_CASE(TestCompareBoundingBoxes3D)
+
+  mesh->computeBoundingBox();
+
+  using Access = receivedPartitionFixture;
```
```suggestion
  using Access = ReceivedPartitionFixture;
```

atotoun replied:
done

BenjaminRodenberg commented on src/partition/tests/fixtures.hpp:
```diff
@@ -0,0 +1,27 @@
+ *
+ * The fixture can be used to call private functions for individual testing. 
+ */
+struct receivedPartitionFixture {
```
```suggestion
struct ReceivedPartitionFixture {
```

atotoun replied:
done

BenjaminRodenberg reviewed (changes requested):
* structs should be capitalized.
* `src/partition/tests/fixtures.hpp` could also go into `src/testing/ReceivePartitionFixture.hpp`. See https://github.com/precice/precice/pull/1050/files#r689567561
